### ISSUE TITLE
[y_groups] Group_SetPlayer is not working with y_hooks

### ISSUE THREAD

**Issue by nikoewe** [pawn-lang/YSI-Includes#533]
I'm trying to make a command permissions module hooking OnGameModeInit and OnPlayerConnect. First, i undefine all commands for the global group on OnGameModeInit, then i create the gUser group an set its commands (for this example, /me). I hook OnPlayerConnect and inside i add the player to gUser group. I made some debug prints and the player IS inside that group, but /me command is not working, returning COMMAND_DENIED value. That's happening when i return Y_HOOKS_CONTINUE_RETURN_1 in OnPlayerConnect, but if i return Y_HOOKS_BREAK_RETURN_1 it works, as well if i use directly the public function.

```pawn
#include <a_samp>
#include <YSI_Players\y_groups>
#include <YSI_Visual\y_commands>
#include <YSI_Coding\y_hooks>

new 
    Group:gUser,
    Group:gOwner;

public e_COMMAND_ERRORS:OnPlayerCommandReceived(playerid, cmdtext[], e_COMMAND_ERRORS:success)
{
    if(success == COMMAND_DENIED)
    {
        SendClientMessage(playerid, -1, "COMMAND DENIED");
        return COMMAND_DENIED;
    }
    return COMMAND_OK;
}
//Owner group unused yet
hook OnGameModeInit()
{
    Group_SetGlobalCommandDefault(UNDEF);
    gUser = Group_Create("gUser");
    gOwner = Group_Create("gOwner");
    Group_AddChild(gOwner, gUser);
    Group_SetCommand(gUser, YCMD:me, ALLOW);
    Command_SetDeniedReturn(true);
    return Y_HOOKS_CONTINUE_RETURN_1;
}

hook OnPlayerConnect(playerid)
{
    Group_SetPlayer(gUser, playerid, true);
    return Y_HOOKS_CONTINUE_RETURN_1;
}


//this should work for gUser
YCMD:me(playerid, params[], help)
{
    SendClientMessage(playerid, -1, "me command");
    return COMMAND_OK;
}


//this should NOT work for gUser
YCMD:do(playerid, params[], help)
{
    SendClientMessage(playerid, -1, "do command");
    return COMMAND_OK;
}

**Comment by Rewgger** [comment]
EDIT: NVM ~1 is to break the chain, 1 is to continue,  but that doesn't work..

(I guess I have solved the issue, scroll down read TLDR; or read everything if you want to know what's happening)

I would like to add something to this as well:

```pawn
#include <a_samp>
#include <YSI_Players\y_groups>
#include <YSI_Visual\y_commands>
#include <YSI_Coding\y_hooks>

new 
    Group:gUser,
    Group:gOwner;

public e_COMMAND_ERRORS:OnPlayerCommandReceived(playerid, cmdtext[], e_COMMAND_ERRORS:success)
{
    if(success == COMMAND_DENIED)
    {
        SendClientMessage(playerid, -1, "No access mate.");
        return COMMAND_DENIED;
    }
    return COMMAND_OK;
}

public OnGameModeInit()
{
    Group_SetGlobalCommandDefault(UNDEF);
    gUser = Group_Create("gUser");
    gOwner = Group_Create("Owner");
    Group_AddChild(gOwner, gUser);
    Group_SetCommand(gUser, YCMD:me, ALLOW);
    Command_SetDeniedReturn(true);
    return 1;
}

public OnPlayerConnect(playerid)
{
    Group_SetPlayer(gUser, playerid, true);
    return 1;
}

YCMD:me(playerid, params[], help)
{
    SendClientMessage(playerid, -1, "It works!");
    return COMMAND_OK;
}

YCMD:do(playerid, params[], help)
{
    SendClientMessage(playerid, -1, "This doesn't work!");
    return COMMAND_OK;
}
```

I have found something really really weird, the above example works, but:

When you hook OnGameModeInit and OnPlayerConnect and return ~1 to both, it also works, but if you return 1, or don't return anything at all, it stops working. Writing all this I actually realized that I answered my own question;
Stop using the macros and actually return ~1 to continue the hook chain, or ~0 to break it. For unknown reasons when you return the macro, it stops working, the only way it works is if you return Y_HOOKS_BREAK_RETURN_1 under OnPlayerConnect.

TLDR; When hooking, instead of returning macro, return either ~1(to continue the hook chain) or ~0(to break the hook chain), otherwise you are going to break it somehow.

**Comment by Rewgger** [comment]
After some debugging and testing I found out that if you use public, you will be able to get it working, with hooking the only way would be if you can make your hook be called AFTER all in YSI, because it's clear that in the library player is kicked out the group when they connect/disconnect, so yea.

**Comment by ghost** [comment]
```
Stop using the macros and actually return ~1 to continue the hook chain, or ~0 to break it. For unknown reasons when you return the macro, it stops working, the only way it works is if you return Y_HOOKS_BREAK_RETURN_1 under OnPlayerConnect.
```
The difference between "return ~1" and "return Y_HOOKS_BREAK_RETURN_1" is that "~1" will be wrapped in brackets (this is from the YSI library):
```c
#define Y_HOOKS_CONTINUE_RETURN_1	(1)		// Continue the hook chain, return 1
#define Y_HOOKS_CONTINUE_RETURN_0	(0)		// Continue the hook chain, return 0
#define Y_HOOKS_BREAK_RETURN_0		(~0)	// Break the hook chain, return 0
#define Y_HOOKS_BREAK_RETURN_1		(~1)	// Break the hook chain, return 1
```

Both OnPlayerConnect() and OnPlayerDisconnect() are hooked in *y_groups_impl.inc*. It is not weird at all that it works when you break the hook: That stops the one in YSI from being called, which you shouldn't do. It's there for a reason, and that would stop all your own hooks that comes after it aswell.

`After some debugging and testing I found out that if you use public, you will be able to get it working`
Off course, because then the player is already initialised in y_groups.

Instead as a workaround you could explicitely hook it to be called first. Y_Less already said he could make the OnPlayerDisconnect() in y_groups be called last (and probably do the same thing for OnPlayerConnect(), but then as first, or atleast as soon as possible, no idea if y_groups depends on anything already), so you'd have to wait for an update. But untill then, make sure your hooks are executed first (and OnPlayerConnect after y_groups' init)

p.s.
`but if you return 1, or don't return anything at all, it stops working`
Not returning anything implies the default return value of that callback. So "not returning anything at all" actually returns "1" (atleast for that callback)
